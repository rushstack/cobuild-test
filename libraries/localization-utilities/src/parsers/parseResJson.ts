// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { JsonFile } from '@rushstack/node-core-library';

import { ILocalizationFile, IParseFileOptions } from '../interfaces';

/**
 * @public
 */
export function parseResJson({ content, ignoreString, filePath }: IParseFileOptions): ILocalizationFile {
  const resjsonFile: Record<string, string> = JsonFile.parseString(content);
  const parsedFile: ILocalizationFile = {};

  const usedComments: Map<string, boolean> = new Map();
  for (const [key, value] of Object.entries(resjsonFile)) {
    if (key.startsWith('_') && key.endsWith('.comment')) {
      if (!usedComments.has(key)) {
        usedComments.set(key, false);
      }
    } else {
      const commentKey: string = `_${key}.comment`;
      const comment: string | undefined = resjsonFile[commentKey];
      usedComments.set(commentKey, true);

      if (!ignoreString?.(filePath, key)) {
        parsedFile[key] = { value, comment };
      }
    }
  }

  const orphanComments: string[] = [];
  for (const [key, used] of usedComments) {
    if (!used) {
      orphanComments.push(key.slice(1, -'.comment'.length));
    }
  }

  if (orphanComments.length > 0) {
    throw new Error(
      'The resjson file is invalid. Comments exist for the following string keys ' +
        `that don't have values: ${orphanComments.join(', ')}.`
    );
  }

  return parsedFile;
}
