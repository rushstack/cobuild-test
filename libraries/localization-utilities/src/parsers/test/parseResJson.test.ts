// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { IgnoreStringFunction } from '../../interfaces';
import { parseResJson } from '../parseResJson';

describe(parseResJson.name, () => {
  it('parses a valid file', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      '_foo.comment': 'A string',
      bar: 'Bar',
      '_bar.comment': 'Another string'
    });

    expect(
      parseResJson({
        content,
        filePath: 'test.resjson'
      })
    ).toMatchSnapshot();
  });

  it('throws on excess comments', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      '_bar.comment': 'A string'
    });

    expect(() =>
      parseResJson({
        content,
        filePath: 'test.resjson'
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('correctly ignores a string', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      '_foo.comment': 'A string',
      bar: 'Bar',
      '_bar.comment': 'Another string'
    });

    const ignoredStringFunction: IgnoreStringFunction = jest
      .fn()
      .mockImplementation(
        (fileName: string, stringName: string) => fileName === 'test.resjson' && stringName === 'bar'
      );

    expect(
      parseResJson({
        content,
        filePath: 'test.resjson',
        ignoreString: ignoredStringFunction
      })
    ).toMatchSnapshot('Loc file');

    expect((ignoredStringFunction as unknown as jest.SpyInstance).mock.calls).toMatchSnapshot(
      'ignoreStrings calls'
    );
  });
});
