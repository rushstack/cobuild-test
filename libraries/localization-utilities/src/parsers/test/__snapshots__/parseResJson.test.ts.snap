// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseResJson correctly ignores a string: Loc file 1`] = `
Object {
  "foo": Object {
    "comment": "A string",
    "value": "Foo",
  },
}
`;

exports[`parseResJson correctly ignores a string: ignoreStrings calls 1`] = `
Array [
  Array [
    "test.resjson",
    "foo",
  ],
  Array [
    "test.resjson",
    "bar",
  ],
]
`;

exports[`parseResJson parses a valid file 1`] = `
Object {
  "bar": Object {
    "comment": "Another string",
    "value": "Bar",
  },
  "foo": Object {
    "comment": "A string",
    "value": "Foo",
  },
}
`;

exports[`parseResJson throws on excess comments 1`] = `"The resjson file is invalid. Comments exist for the following string keys that don't have values: bar."`;
